File: apps/mobile/src/utils/rateLimit.ts
/**
 * rateLimit.ts — sliding-window rate limiter for client-side calls.
 *
 * IMPORTANT: this is a UX guard, NOT a security boundary. Real rate limiting
 * MUST live server-side (Postgres functions, Supabase Edge Functions, or a
 * gateway). Anything that runs in JavaScript on the user's device can be
 * bypassed by a determined attacker.
 *
 * Use this for:
 *   - Stopping users from accidentally double-tapping a "post" button.
 *   - Limiting expensive search queries to avoid hammering the backend.
 *   - Telegraphing rate limits to users with a friendly error before they hit
 *     a server-side 429.
 *
 * Don't use this for:
 *   - Preventing abuse (e.g., spam, scraping). Use Supabase RLS + a server-
 *     side limiter (CREATE FUNCTION + a counter table).
 *   - Auth flows — Supabase Auth has its own rate limits.
 *
 * Pair with server-side enforcement.
 */

export class RateLimitError extends Error {
  public readonly retryAfterMs: number;

  constructor(retryAfterMs: number) {
    super(`Rate limit hit. Try again in ${Math.ceil(retryAfterMs / 1000)}s.`);
    this.name = 'RateLimitError';
    this.retryAfterMs = retryAfterMs;
  }
}

export interface RateLimiter {
  /** Records a call. Throws RateLimitError if the window is full. */
  check: () => void;
  /** True if the next `check()` would succeed. Doesn't record a call. */
  isAllowed: () => boolean;
  /** Reset the window (e.g., after a successful retry). */
  reset: () => void;
}

/**
 * Build a sliding-window rate limiter.
 *
 *   const limiter = createRateLimiter({ maxCalls: 5, windowMs: 60_000 });
 *   try { limiter.check(); await postSomething(); }
 *   catch (e) { if (e instanceof RateLimitError) showToast(e.message); }
 */
export function createRateLimiter(options: {
  readonly maxCalls: number;
  readonly windowMs: number;
}): RateLimiter {
  const { maxCalls, windowMs } = options;
  if (maxCalls < 1) {
    throw new Error('maxCalls must be at least 1');
  }
  if (windowMs < 1) {
    throw new Error('windowMs must be at least 1');
  }

  let timestamps: number[] = [];

  function evict(now: number): void {
    const cutoff = now - windowMs;
    while (timestamps.length > 0 && timestamps[0] !== undefined && timestamps[0] < cutoff) {
      timestamps.shift();
    }
  }

  return {
    check(): void {
      const now = Date.now();
      evict(now);
      if (timestamps.length >= maxCalls) {
        const oldest = timestamps[0] ?? now;
        throw new RateLimitError(oldest + windowMs - now);
      }
      timestamps.push(now);
    },

    isAllowed(): boolean {
      const now = Date.now();
      evict(now);
      return timestamps.length < maxCalls;
    },

    reset(): void {
      timestamps = [];
    },
  };
}

File: ios/Venn/Utils/RateLimit.swift
import Foundation

/// Sliding-window rate limiter for client-side UX feedback only.
///
/// IMPORTANT — this is not a security boundary. Real rate limiting MUST live
/// server-side (Postgres `rl_check` function called by Edge Functions / RPCs;
/// see `docs/CODING_STANDARDS.md` § Rate limiting). Anything that runs in the
/// app on the user's device can be bypassed by a determined attacker.
///
/// Use cases:
///   - Stop accidental double-taps of "post" buttons.
///   - Throttle expensive search queries to avoid hammering the backend.
///   - Show a friendly "slow down" toast before the server returns 429.
///
/// Modelled as an `actor` so concurrent callers across `Task`s see consistent
/// state under `SWIFT_STRICT_CONCURRENCY=complete`.
actor RateLimiter {
    /// Configuration: how many calls in what window.
    struct Limit: Equatable {
        let maxCalls: Int
        let windowSeconds: TimeInterval

        init(maxCalls: Int, windowSeconds: TimeInterval) {
            precondition(maxCalls >= 1, "maxCalls must be at least 1")
            precondition(windowSeconds > 0, "windowSeconds must be positive")
            self.maxCalls = maxCalls
            self.windowSeconds = windowSeconds
        }
    }

    /// Thrown by `check()` when the window is full.
    struct LimitExceeded: Error, Equatable {
        /// Seconds until the *oldest* recorded call ages out and frees a slot.
        let retryAfter: TimeInterval
    }

    private let limit: Limit
    private var calls: [Date] = []

    init(limit: Limit) {
        self.limit = limit
    }

    /// Records a call against the limiter. Throws `LimitExceeded` if the
    /// window is already full. The `now` parameter is exposed for tests so
    /// they can simulate time without relying on `Date()`.
    func check(now: Date = Date()) throws {
        evict(before: now.addingTimeInterval(-limit.windowSeconds))
        if calls.count >= limit.maxCalls {
            let oldest = calls.first ?? now
            let retryAfter = limit.windowSeconds - now.timeIntervalSince(oldest)
            throw LimitExceeded(retryAfter: max(retryAfter, 0))
        }
        calls.append(now)
    }

    /// Non-mutating: would the next `check()` succeed right now?
    func isAllowed(now: Date = Date()) -> Bool {
        evict(before: now.addingTimeInterval(-limit.windowSeconds))
        return calls.count < limit.maxCalls
    }

    /// Forget every recorded call. Use after a successful retry, on sign-out,
    /// or when the user changes context (e.g., switches profiles).
    func reset() {
        calls.removeAll()
    }

    private func evict(before cutoff: Date) {
        calls.removeAll { $0 < cutoff }
    }
}
